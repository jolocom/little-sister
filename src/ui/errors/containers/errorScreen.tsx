import React, { useEffect } from 'react'
import { BackHandler, StyleSheet, Text, View } from 'react-native'
import { JolocomButton, Wrapper } from '../../structure'
import { NavigationScreenProp, NavigationState } from 'react-navigation'
import { centeredText, fontMain, fontMedium } from '../../../styles/typography'
import { Colors, Typefaces } from '../../../styles'
import { BlueCircleImage, RedCircleImage } from '../../../resources'
import { BP } from '../../../styles/breakpoints'
import { connect } from 'react-redux'
import { ThunkDispatch } from '../../../store'
import { navigationActions } from '../../../actions'
import { routeList } from '../../../routeList'

const styles = StyleSheet.create({
  wrapper: {
    backgroundColor: Colors.overflowBlack,
    justifyContent: 'space-between',
  },
  imageWrapper: {
    marginTop: '15%',
    width: '65%',
    height: '54%',
  },
  title: {
    ...Typefaces.title2,
    ...centeredText,
    fontFamily: fontMedium,
    color: Colors.sandLight090,
    paddingHorizontal: 10,
  },
  message: {
    ...centeredText,
    fontFamily: fontMain,
    lineHeight: BP({
      small: 20,
      medium: 20,
      large: 24,
    }),
    fontSize: BP({
      small: 18,
      medium: 18,
      large: 20,
    }),
    marginTop: 13,
    paddingHorizontal: 26,
    color: Colors.sandLight080,
  },
  interactButton: {
    marginTop: 14,
    borderColor: 'white',
    borderWidth: 0.5,
  },
  buttonSection: {
    flex: 0.3,
    width: '100%',
    paddingHorizontal: 20,
    justifyContent: 'center',
  },
  contentSection: {
    flex: 0.7,
    width: '100%',
    alignItems: 'center',
  },
})

export enum ImageType {
  Red = 'red',
  Blue = 'blue',
}

export interface ErrorScreenParams {
  title: string
  message: string
  image?: ImageType
  interact?: {
    label: string
    onInteract: () => void | Promise<void>
  }
  dismiss: {
    label: string
    onDismiss: () => void | Promise<void>
  }
}

interface Props extends ReturnType<typeof mapDispatchToProps> {
  navigation: NavigationScreenProp<NavigationState, ErrorScreenParams>
}

const ErrorScreenContainer: React.FC<Props> = props => {
  const {
    resetNavigation,
    navigation: {
      state: {
        params: { title, message, interact, dismiss, image },
      },
    },
  } = props

  useEffect(() => {
    BackHandler.addEventListener('hardwareBackPress', onBackButtonPress)

    return () => {
      BackHandler.removeEventListener('hardwareBackPress', onBackButtonPress)
    }
  }, [])

  const onBackButtonPress = (): boolean => {
    resetNavigation()
    return true
  }

  return (
    <Wrapper style={styles.wrapper}>
      <View style={styles.contentSection}>
        <View style={styles.imageWrapper}>
          {image === ImageType.Red ? <RedCircleImage /> : <BlueCircleImage />}
        </View>
        <Text style={styles.title}>{title}</Text>
        <Text style={styles.message}>{message}</Text>
      </View>
      <View style={styles.buttonSection}>
        <JolocomButton
          onPress={dismiss.onDismiss}
          transparent
          text={dismiss.label}
        />
        {interact && (
          <JolocomButton
            containerStyle={styles.interactButton}
            transparent
            onPress={interact.onInteract}
            text={interact.label}
          />
        )}
      </View>
    </Wrapper>
  )
}

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  resetNavigation: () =>
    dispatch(navigationActions.navigate({ routeName: routeList.AppInit })),
})

export const ErrorScreen = connect(
  null,
  mapDispatchToProps,
)(ErrorScreenContainer)
