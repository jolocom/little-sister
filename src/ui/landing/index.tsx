export { Landing } from './containers/landing'
