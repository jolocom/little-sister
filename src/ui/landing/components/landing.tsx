import React from 'react'
import { StyleSheet, Dimensions, View, Text, Animated } from 'react-native'
import { Wrapper, OnlineJolocomButton } from 'src/ui/structure'
import I18n from 'src/locales/i18n'
import strings from 'src/locales/strings'
import { landingSlides, Slide } from './landingSlides'
import Carousel, {
  CarouselProps,
  getInputRangeFromIndexes,
} from 'react-native-snap-carousel'
import { Typography, Colors, Spacing, Buttons } from 'src/styles'

interface Props {
  handleGetStarted: () => void
  handleRecover: () => void
}

const viewWidth = Dimensions.get('window').width

const styles = StyleSheet.create({
  mainContainer: {
    backgroundColor: Colors.blackMain,
  },
  carouselSlide: {
    width: '100%',
    height: '100%',
    alignItems: 'center',
  },
  carouselTextContainer: {
    marginTop: viewWidth > 360 ? Spacing['4XL'] : Spacing.XXL,
    position: 'absolute',
    paddingHorizontal: '5%',
  },
  header: {
    ...Typography.mainText,
    textAlign: 'center',
    color: Colors.sandLight,
  },
  message: {
    ...Typography.subMainText,
    textAlign: 'center',
    color: Colors.sandLight080,
    lineHeight: Typography.subMainText.fontSize + 4,
    marginTop: Spacing.SM,
  },
  activeDotStyle: {
    width: 8,
    height: 8,
    backgroundColor: Colors.dotColorActive,
  },
  inactiveDotStyle: {
    width: 8,
    height: 8,
    backgroundColor: Colors.dotColorInactive,
  },
  bottomSection: {
    position: 'absolute',
    width: '85%',
    bottom: '5%',
    alignItems: 'center',
  },
  mainButtonContainer: {
    ...Buttons.buttonStandardContainer,
    alignSelf: 'stretch',
  },
  mainButtonText: {
    ...Buttons.buttonStandardText,
  },
  recoverButtonContainer: {
    ...Buttons.buttonStandardContainer,
    backgroundColor: 'transparent',
    marginTop: Spacing.XS,
    alignSelf: 'stretch',
  },
  recoverButtonText: {
    ...Buttons.buttonStandardText,
    fontSize: Typography.textSM,
  },
})

export class LandingComponent extends React.Component<Props> {
  // https://github.com/archriss/react-native-snap-carousel/blob/master/doc/CUSTOM_INTERPOLATIONS.md
  // 0 is the current slide, and we want there to be an animated fade in/out
  private scrollInterpolator(index: number, carouselProps: CarouselProps<any>) {
    const range = [1, 0, -1]
    const inputRange = getInputRangeFromIndexes(range, index, carouselProps)
    return { inputRange, outputRange: range }
  }

  private animatedStyles(
    index: number,
    animatedValue: Animated.AnimatedValue,
    carouselProps: CarouselProps<any>,
  ) {
    return {
      opacity: animatedValue.interpolate({
        inputRange: [-1, 0, 1],
        outputRange: [-1.5, 1, -1.5],
      }),
    }
  }

  private renderItem = ({ item }: { item: Slide }) => {
    const { bgImage, title, infoText } = item
    return (
      <View style={styles.carouselSlide}>
        {bgImage}
        <View style={styles.carouselTextContainer}>
          <Text style={styles.header}>{title}</Text>
          <Text style={styles.message}>{infoText}</Text>
        </View>
      </View>
    )
  }

  public render() {
    return (
      <Wrapper style={styles.mainContainer}>
        <Carousel
          testID="landingCarousel"
          data={landingSlides}
          renderItem={this.renderItem}
          lockScrollWhileSnapping
          lockScrollTimeoutDuration={1000}
          loop
          autoplay
          autoplayDelay={5000}
          autoplayInterval={5000}
          sliderWidth={viewWidth}
          itemWidth={viewWidth}
          layout={'default'}
          scrollInterpolator={this.scrollInterpolator}
          /** @TODO Fix typing? */
          // @ts-ignore
          slideInterpolatedStyle={this.animatedStyles}
        />
        <View style={styles.bottomSection}>
          <OnlineJolocomButton
            testID="getStarted"
            containerStyle={{ width: '100%' }}
            onPress={this.props.handleGetStarted}
            text={I18n.t(strings.GET_STARTED)}
          />
          <OnlineJolocomButton
            testID="recoverIdentity"
            containerStyle={{
              width: '100%',
              marginTop: 12,
              height: 48,
            }}
            onPress={this.props.handleRecover}
            text={I18n.t(strings.RECOVER_IDENTITY)}
            transparent={true}
          />
        </View>
      </Wrapper>
    )
  }
}
