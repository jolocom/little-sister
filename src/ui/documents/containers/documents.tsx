import React from 'react'
import { connect } from 'react-redux'
import { Animated, ScrollView, StyleSheet, Text, View } from 'react-native'

import { openDocumentDetails } from 'src/actions/documents'
import { DecoratedClaims } from 'src/reducers/account'
import { RootState } from 'src/reducers'
import { getDocumentClaims } from 'src/utils/filterDocuments'
import { ThunkDispatch } from 'src/store'

import I18n from 'src/locales/i18n'
import { filters } from 'src/lib/filterDecoratedClaims'

import { DocumentsCarousel } from '../components/documentsCarousel'
import { DocumentsList } from '../components/documentsList'
import { DocumentViewToggle } from '../components/documentViewToggle'
import strings from '../../../locales/strings'
import { Typography, Colors } from 'src/styles'
import { Wrapper } from '../../structure'

interface Props
  extends ReturnType<typeof mapDispatchToProps>,
    ReturnType<typeof mapStateToProps> {}

interface State {
  activeDocumentIndex: number
  showingValid: boolean
}

const styles = StyleSheet.create({
  emptyDocumentsText: {
    ...Typography.mainText,
    ...Typography.centeredText,
    color: Colors.greyLight,
    paddingHorizontal: '5%',
  },
})

export class DocumentsContainer extends React.Component<Props, State> {
  public ScrollViewRef: ScrollView | null = null

  public state = {
    activeDocumentIndex: 0,
    showingValid: true,
  }

  private scrollToTop(): void {
    if (this.ScrollViewRef) {
      this.ScrollViewRef.scrollTo({ x: 0, y: 0, animated: true })
    }
  }

  private onActiveIndexChange(index: number): void {
    this.setState({ activeDocumentIndex: index })
    this.scrollToTop()
  }

  private handleToggle = () => {
    this.setState((prevState: State) => ({
      showingValid: !prevState.showingValid,
      activeDocumentIndex: 0,
    }))
    this.scrollToTop()
  }

  public render(): JSX.Element {
    const { openDocumentDetails, decoratedCredentials } = this.props
    const documents = getDocumentClaims(decoratedCredentials['Other'])
    const docFilter = this.state.showingValid
      ? filters.filterByValid
      : filters.filterByExpired
    const displayedDocs = docFilter(documents)
    const isEmpty = displayedDocs.length === 0
    const otherIsEmpty = displayedDocs.length === documents.length

    if (isEmpty) {
      return (
        <Wrapper centered>
          <Text style={styles.emptyDocumentsText}>
            {I18n.t(strings.NO_DOCUMENTS_TO_SEE_HERE) + '...'}
          </Text>
        </Wrapper>
      )
    }

    return (
      <Wrapper>
        <Animated.View style={{ flex: 1 }}>
          {!otherIsEmpty && (
            <DocumentViewToggle
              showingValid={this.state.showingValid}
              onTouch={this.handleToggle}
            />
          )}
          <ScrollView
            // to scroll to top upon changing card
            ref={ref => (this.ScrollViewRef = ref)}
            contentContainerStyle={
              isEmpty ? { flex: 1 } : { paddingBottom: 110 }
            }
            scrollEnabled={!isEmpty}
          >
            {this.state.showingValid ? (
              <DocumentsCarousel
                documents={displayedDocs}
                activeIndex={this.state.activeDocumentIndex}
                onActiveIndexChange={this.onActiveIndexChange.bind(this)}
              />
            ) : (
              <DocumentsList
                documents={displayedDocs}
                onDocumentPress={openDocumentDetails}
              />
            )}
          </ScrollView>
        </Animated.View>
      </Wrapper>
    )
  }
}

const mapStateToProps = ({
  account: {
    claims: { decoratedCredentials },
  },
}: RootState) => ({
  decoratedCredentials,
})

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  openDocumentDetails: (doc: DecoratedClaims) =>
    dispatch(openDocumentDetails(doc)),
})

export const Documents = connect(
  mapStateToProps,
  mapDispatchToProps,
)(DocumentsContainer)
