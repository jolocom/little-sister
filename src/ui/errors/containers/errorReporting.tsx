import React, { useState } from 'react'
import { connect } from 'react-redux'
import { NavigationScreenProp, NavigationState } from 'react-navigation'
import { AppError, reportError } from '../../../lib/errors'
import { ThunkDispatch } from '../../../store'
import { routeList } from '../../../routeList'
import { navigationActions } from '../../../actions'
import { Emoji, EmojiSection } from '../components/emojiSection'
import { JolocomButton, Wrapper } from '../../structure'
import { ScrollView } from 'react-native'
import { ChooseIssueSection } from '../components/chooseIssueSection'
import { DescriptionSection } from '../components/descriptionSection'
import { ContactSection } from '../components/contactSection'
import I18n from '../../../locales/i18n'
import strings from '../../../locales/strings'
import { SectionWrapper } from '../components/sectionWrapper'
import { NavigationSection } from '../components/navigationSection'

interface PaymentNavigationParams {
  error?: AppError | Error
  previousScreen: routeList
}

interface Props extends ReturnType<typeof mapDispatchToProps> {
  navigation: NavigationScreenProp<NavigationState, PaymentNavigationParams>
}

export enum Inputs {
  None,
  Dropdown,
  Description,
  Contact,
}

const ErrorReportingContainer = (props: Props) => {
  const { navigateToScreen, navigation } = props

  const [pickedIssue, setIssue] = useState('')
  const [description, setDescription] = useState('')
  const [contact, setContact] = useState('')
  const [currentInput, setInput] = useState(Inputs.None)
  const [toggleState, setToggle] = useState(false)
  const [selectedEmoji, setEmoji] = useState(Emoji.Empty)

  // NOTE send previous screen as props, since not all
  // error reports will have error objects

  // const { previousScreen } = navigation.state.params
  // const isBackButton = previousScreen === routeList.Settings
  const isBackButton = false

  // NOTE error reports can show up without an error,
  // hence a navigation prop (navigateTo) would be better than in the error object
  const navigateBack = () => {
    const error = props.navigation?.state?.params?.error
    const previousScreen = props.navigation?.state.params?.previousScreen
    if (error instanceof AppError) {
      navigateToScreen(error.navigateTo)
    } else {
      navigateToScreen(previousScreen || routeList.AppInit)
    }
  }

  const onSubmitReport = () => {
    const error = props.navigation?.state?.params?.error

    const userReport = {
      userError: pickedIssue,
      userDescription: description,
      userContact: contact,
      sendPrivateData: toggleState,
    }

    if (navigation) {
      reportError({ ...userReport, error })
      navigateBack()
    }
  }

  return (
    <Wrapper secondaryDark>
      <NavigationSection
        onNavigation={navigateBack}
        isBackButton={isBackButton}
      />
      <ScrollView overScrollMode="never">
        <SectionWrapper
          title={I18n.t(strings.TELL_US_THE_PROBLEM)}
          style={{ marginTop: 14 }}>
          <ChooseIssueSection
            currentInput={currentInput}
            pickedIssue={pickedIssue}
            onIssuePick={setIssue}
            setInput={setInput}
          />
        </SectionWrapper>
        <SectionWrapper title={I18n.t(strings.CAN_YOU_BE_MORE_SPECIFIC)}>
          <DescriptionSection
            currentInput={currentInput}
            setInput={setInput}
            setDescription={setDescription}
            toggleState={toggleState}
            setToggle={setToggle}
            description={description}
          />
        </SectionWrapper>
        <SectionWrapper title={I18n.t(strings.WANT_TO_GET_IN_TOUCH)}>
          <ContactSection
            onContactInput={setContact}
            currentInput={currentInput}
            setInput={setInput}
            contactValue={contact}
          />
        </SectionWrapper>
        <SectionWrapper title={I18n.t(strings.RATE_THE_ISSUE)}>
          <EmojiSection selectedEmoji={selectedEmoji} setEmoji={setEmoji} />
        </SectionWrapper>
        <JolocomButton
          onPress={onSubmitReport}
          containerStyle={{
            marginTop: 45,
            marginBottom: 66,
            marginHorizontal: 20,
            height: 56,
          }}
          text={I18n.t(strings.SUBMIT_REPORT)}
        />
      </ScrollView>
    </Wrapper>
  )
}

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  navigateToScreen: (screen: routeList) =>
    dispatch(navigationActions.navigate({ routeName: screen })),
})

export const ErrorReporting = connect(
  null,
  mapDispatchToProps,
)(ErrorReportingContainer)
