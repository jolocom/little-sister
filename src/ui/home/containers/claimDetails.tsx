import React from 'react'
import { ClaimDetailsComponent } from 'src/ui/home/components/claimDetails'
import { connect } from 'react-redux'
import { RootState } from 'src/reducers/'
import { accountActions } from 'src/actions'
import { ClaimsState } from 'src/reducers/account'
import Immutable from 'immutable'

interface ConnectProps {
  claims: ClaimsState
  saveClaim: () => void
  handleClaimInput: (fieldValue: string, fieldName: string) => void
}

interface Props extends ConnectProps {}

interface State {}

export class ClaimDetailsContainer extends React.Component<Props, State> {
  render() {
    return (
      <ClaimDetailsComponent
        saveClaim={this.props.saveClaim}
        handleClaimInput={this.props.handleClaimInput}
        selectedClaim={this.props.claims.selected}
      />
    )
  }
}

const mapStateToProps = (state: RootState) => {
  const claims = Immutable.fromJS(state.account.claims)
  return {
    claims: claims.toJS(),
  }
}

const mapDispatchToProps = (dispatch: Function) => ({
  saveClaim: () => {
    dispatch(accountActions.saveClaim())
  },
  handleClaimInput: (fieldValue: string, fieldName: string) => {
    dispatch(accountActions.handleClaimInput(fieldValue, fieldName))
  },
})

export const ClaimDetails = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ClaimDetailsContainer)
