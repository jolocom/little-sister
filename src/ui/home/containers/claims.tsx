import React from 'react'
import { connect } from 'react-redux'
import { View } from 'react-native'

import { CredentialOverview } from '../components/credentialOverview'
import { accountActions, ssoActions, navigationActions } from 'src/actions'
import { ClaimsState } from 'src/reducers/account'
import { DecoratedClaims } from 'src/reducers/account/'
import { routeList } from 'src/routeList'

interface ConnectProps {
  setClaimsForDid: () => void
  toggleLoading: (val: boolean) => void
  openClaimDetails: (claim: DecoratedClaims) => void
  openScanner: () => void
  parseJWT: (jwt: string) => void
  did: string
  claims: ClaimsState
  loading: boolean
}

interface Props extends ConnectProps {}

export class ClaimsContainer extends React.Component<Props> {
  componentWillMount() {
    this.props.setClaimsForDid()
  }

  render() {
    return (
      <View style={{ flex: 1 }}>
        <CredentialOverview
          did={this.props.did}
          claimsState={this.props.claims}
          loading={!!this.props.claims.loading}
          onEdit={this.props.openClaimDetails}
        />
      </View>
    )
  }
}

// TODO nicer pattern for accessing state, perhaps immer or something easier to Type
const mapStateToProps = (state: any) => ({
  did: state.account.did.toJS().did,
  claims: state.account.claims.toJS(),
  loading: state.account.loading.toJS().loading,
})

const mapDispatchToProps = (dispatch: Function) => ({
  parseJWT: (jwt: string) => dispatch(ssoActions.parseJWT(jwt)),
  openClaimDetails: (claim: DecoratedClaims) =>
    dispatch(accountActions.openClaimDetails(claim)),
  setClaimsForDid: () => dispatch(accountActions.setClaimsForDid()),
  toggleLoading: (val: boolean) => dispatch(accountActions.toggleLoading(val)),
  openScanner: () =>
    dispatch(
      navigationActions.navigate({ routeName: routeList.QRCodeScanner }),
    ),
})

export const Claims = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ClaimsContainer)
