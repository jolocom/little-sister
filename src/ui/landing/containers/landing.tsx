import React from 'react'
import { connect } from 'react-redux'
import { LandingComponent } from 'src/ui/landing/components/landing'
import { navigationActions } from 'src/actions/'
import { ThunkDispatch } from 'src/store'
import { StatusBar } from 'react-native'
import { routeList } from '../../../routeList'
import { withInternet } from '../../../actions/modifiers'

interface Props extends ReturnType<typeof mapDispatchToProps> {}

export const LandingContainer = (props: Props): JSX.Element => {
  return (
    <React.Fragment>
      <StatusBar barStyle="light-content" />
      <LandingComponent
        handleGetStarted={props.getStarted}
        handleRecover={props.recoverIdentity}
      />
    </React.Fragment>
  )
}

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  getStarted: () =>
    dispatch(
      withInternet(
        navigationActions.navigate({
          routeName: routeList.Entropy,
        }),
      ),
    ),
  recoverIdentity: () =>
    dispatch(
      withInternet(
        navigationActions.navigate({
          routeName: routeList.InputSeedPhrase,
        }),
      ),
    ),
})

export const Landing = connect(
  null,
  mapDispatchToProps,
)(LandingContainer)
