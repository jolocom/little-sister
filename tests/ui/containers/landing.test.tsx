import React from 'react'
import { LandingContainer } from 'src/ui/landing/containers/landing'
import { shallow } from 'enzyme'

describe('landing container', () => {
  it('mounts correctly and matches snapshot', () => {
    const props = {
      getStarted: jest.fn(),
      recoverIdentity: jest.fn(),
    }

    const rendered = shallow(<LandingContainer {...props} />)
    expect(rendered).toMatchSnapshot()
    expect(props.getStarted).not.toHaveBeenCalled()
  })
})
