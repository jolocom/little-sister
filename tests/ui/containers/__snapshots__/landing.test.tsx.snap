// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`landing container mounts correctly and matches snapshot 1`] = `
<LandingComponent
  handleButtonTap={[Function]}
/>
`;
